tests/components/songpal/test_media_player.py

```diff
@@ -0,0 +1,287 @@
+
+
+async def test_websocket_events(hass):
+    """Test websocket events."""
+    mocked_device = _create_mocked_device()
+    entry = MockConfigEntry(domain=songpal.DOMAIN, data=CONF_DATA)
+    entry.add_to_hass(hass)
+
+    with _patch_media_player_device(mocked_device):
+        await hass.config_entries.async_setup(entry.entry_id)
+        await hass.async_block_till_done()
+
+    mocked_device.listen_notifications.assert_called_once()
+    assert mocked_device.on_notification.call_count == 4
+
+    notification_callbacks = mocked_device.notification_callbacks
+
+    volume_change = MagicMock()
+    volume_change.mute = True
+    volume_change.volume = 20
+    await notification_callbacks[VolumeChange](volume_change)
+    attributes = _get_attributes(hass)
+    assert attributes["is_volume_muted"] is True
+    assert attributes["volume_level"] == 0.2
+
+    content_change = MagicMock()
+    content_change.is_input = False
+    content_change.uri = "uri1"
+    await notification_callbacks[ContentChange](content_change)
+    assert _get_attributes(hass)["source"] == "title2"
+    content_change.is_input = True
+    await notification_callbacks[ContentChange](content_change)
+    assert _get_attributes(hass)["source"] == "title1"
+
+    power_change = MagicMock()
+    power_change.status = False
+    await notification_callbacks[PowerChange](power_change)
+    assert hass.states.get(ENTITY_ID).state == STATE_OFF
+
+    connect_change = MagicMock()
+    connect_change.exception = "disconnected"
+    with patch("asyncio.sleep", side_effect=AsyncMock()) as sleep:
+        await notification_callbacks[ConnectChange](connect_change)
+    sleep.assert_called_once()
+    mocked_device.clear_notification_callbacks.assert_called_once()
+
+
+async def test_disconnected(hass):
+    """Test disconnected behavior."""
+    mocked_device = _create_mocked_device()
+    entry = MockConfigEntry(domain=songpal.DOMAIN, data=CONF_DATA)
+    entry.add_to_hass(hass)
+
+    with _patch_media_player_device(mocked_device):
+        await hass.config_entries.async_setup(entry.entry_id)
+        await hass.async_block_till_done()
+
+    logger_warning = MagicMock()
+    logger_error = MagicMock()
+    connect_change = MagicMock()
+    connect_change.exception = "disconnected"
+    type(mocked_device).get_supported_methods = AsyncMock(
+        side_effect=SongpalException("disconnected")
+    )
+    notification_callbacks = mocked_device.notification_callbacks
+
+    async def _mocked_sleep(delay):
+        if delay == 10:  # first sleep
+            logger_warning.assert_called_once()
+            logger_warning.reset_mock()
+        elif delay == 20:  # second sleep
+            logger_warning.assert_not_called()
+            type(mocked_device).get_supported_methods = AsyncMock()
+        else:
+            assert False  # should never be here
+
+    with patch("asyncio.sleep", side_effect=_mocked_sleep), patch(
```

Comment:
Don't patch sleep. It's too aggressive. It's ok to patch the time we sleep. If we break out the start delay time to a constant it's easy to patch that to 0. Then set the side effect of `get_supported_methods` to a list that raises in the first item and succeeds in the second item.